src/machine/aot/aot.x64.c

```diff
@@ -0,0 +1,1284 @@
+      ret = aot_mov_pc_internal(context, (condition.value.i == 1) ? true_value : false_value);
+      if (ret != DASM_S_OK) { return ret; }
+      break;
+    case AOT_TAG_X64_REGISTER:
+      | cmp Rq(condition.value.x64_reg), 1
+      | jne >1
+      ret = aot_mov_pc_internal(context, true_value);
+      if (ret != DASM_S_OK) { return ret; }
+      |1:
+      ret = aot_mov_pc_internal(context, false_value);
+      if (ret != DASM_S_OK) { return ret; }
+      break;
+  }
+
+  return DASM_S_OK;
+}
+
+int aot_memory_write(AotContext* context, AotValue address, AotValue v, uint32_t size)
+{
+  int ret;
+  dasm_State** Dst = &context->d;
+
+  ret = aot_mov_x64(context, X64_RAX, address);
+  if (ret != DASM_S_OK) { return ret; }
+
+  | mov rcx, rax
+  | shr rcx, CKB_VM_ASM_RISCV_PAGE_SHIFTS
+   /*
+    * Test if the page stored in rcx is out of bound, and if the page has
+    * correct write permissions
+    */
+  | cmp rcx, CKB_VM_ASM_RISCV_PAGES
+  | jae >2
+  | lea rdx, machine->flags
+  | movzx edx, byte [rdx+rcx]
+  | and edx, CKB_VM_ASM_MEMORY_FLAG_WXORX_BIT
+  | cmp edx, CKB_VM_ASM_MEMORY_FLAG_WRITABLE
+  | jne >3
+  /* Check if the write spans to a second memory page */
+  | mov rdx, rax
+  | add rdx, size
+  | sub rdx, 1
+  | shr rdx, CKB_VM_ASM_RISCV_PAGE_SHIFTS
+  | add rcx, 1
+  | cmp rcx, rdx
+  | jne >1
+   /*
+    * Test if the page stored in rcx is out of bound, and if the page has
+    * correct write permissions
+    */
+  | cmp rcx, CKB_VM_ASM_RISCV_PAGES
+  | jae >2
+  | lea rdx, machine->flags
+  | movzx edx, byte [rdx+rcx]
+  | and edx, CKB_VM_ASM_MEMORY_FLAG_WXORX_BIT
+  | cmp edx, CKB_VM_ASM_MEMORY_FLAG_WRITABLE
+  | jne >3
+  |1:
+  | lea rdx, machine->memory
+  ret = aot_mov_x64(context, X64_RCX, v);
+  if (ret != DASM_S_OK) { return ret; }
+  switch (size) {
+    case 1:
+      | mov byte [rdx+rax], cl
+      break;
+    case 2:
+      | mov word [rdx+rax], cx
+      break;
+    case 4:
+      | mov dword [rdx+rax], ecx
+      break;
+    case 8:
+      | mov qword [rdx+rax], rcx
+      break;
+    default:
+      return ERROR_INVALID_MEMORY_SIZE;
+  }
+  | jmp >4
+  |2:
+  ret = aot_exit(context, CKB_VM_ASM_RET_OUT_OF_BOUND);
+  if (ret != DASM_S_OK) { return ret; }
+  |3:
+  ret = aot_exit(context, CKB_VM_ASM_RET_INVALID_PERMISSION);
+  if (ret != DASM_S_OK) { return ret; }
+  |4:
+
+  return DASM_S_OK;
+}
+
+int aot_memory_read(AotContext* context, uint32_t target, AotValue address, uint32_t size)
+{
+  uint32_t loc1;
+  int ret;
+  dasm_State** Dst = &context->d;
+
+  ret = aot_mov_x64(context, X64_RAX, address);
+  if (ret != DASM_S_OK) { return ret; }
+
+  | mov rdx, rax
+  | cmp rdx, CKB_VM_ASM_RISCV_MAX_MEMORY
+  | jae >1
+  | add rdx, size
+  | cmp rdx, CKB_VM_ASM_RISCV_MAX_MEMORY
+  | jae >1
+  | lea rdx, machine->memory
+  switch (size) {
+    case 1:
+      | movzx rcx, byte [rdx+rax]
+      break;
+    case 2:
+      | movzx rcx, word [rdx+rax]
+      break;
+    case 4:
+      | mov ecx, dword [rdx+rax]
```

Comment:
why not use rcx like other case？

Reply:
In x64 assembly when writing to `ecx`, the higher 32-bit of `rcx` will automatically be cleared. So using `mov ecx, dword [rdx+rax]` here has the same effect as `mov rcx, dword [rdx+rax]`, but with the benefit of one less bytes in the encoded assembly, which is nicer to the CPU frontend decoder.

That being said, we should change the code to use `ecx` in all cases here except when we are writing all 64 bits.